File: solution/disjoint_set/16562/main.cpp
// Authored by : tony9402
// Co-authored by : -
// Link : http://boj.kr/caef773ebb994d6e854c390d5fe22057
#include<bits/stdc++.h>

using namespace std;

int uf[10101], arr[10101];

int find(int x) {
    if(uf[x] < 0) return x;
    return uf[x] = find(uf[x]);
}

bool merge(int a, int b) {
    a = find(a);
    b = find(b);
    if(a == b) return false;
    uf[b] = a;
    arr[a] = min(arr[a], arr[b]);
    arr[b] = 0;
    return true;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n, m, k; cin >> n >> m >> k;
    for(int i=1;i<=n;i++) {
        uf[i] = -1;
        cin >> arr[i];
    }
    for(int i=0;i<m;i++) {
        int a, b; cin >> a >> b;
        merge(a, b);
    }
    int answer = 0;
    for(int i=1;i<=n;i++) answer += arr[i];
    if(answer > k) cout << "Oh no";
    else cout << answer;

    return 0;
}

File: solution/disjoint_set/16562/main.swift
# Authored by : iyeaaa
# Co-authored by : -
# Link : http://boj.kr/a5ff34cf2dc845f19b2742d7ab1bc7f9

var input = readLine()!.split{$0==" "}.map{Int(String($0))!}
let (n, m) = (input[0], input[1]); var k = input[2]
let cost = [0] + readLine()!.split{$0==" "}.map{Int(String($0))!}
var friend = Array(0...n)
var minCost = cost

for _ in 0..<m {
    input = readLine()!.split{$0==" "}.map{Int(String($0))!}
    merge(input[0], input[1])
}

var sum = 0
for i in 1...n {
    if isRoot(i) {
        sum += minCost[i]
        k -= minCost[i]
    }
}

print( k < 0 ? "Oh no" : sum )

func find(_ x: Int) -> Int {
    if x == friend[x] { return x }
    friend[x] = find(friend[x])
    return friend[x]
}
func merge(_ x: Int, _ y: Int) {
    let (x, y) = (find(x), find(y))
    if x == y { return }
    friend[x] = y
    if minCost[x] > minCost[y] { minCost[x] = minCost[y] }
    else { minCost[y] = minCost[x] }
}
func isUnion(_ x: Int, _ y: Int) -> Bool {
    find(x) == find(y)
}
func isRoot(_ x: Int) -> Bool {
    x == friend[x]
}
